### PR TITLE
[#1235] Server switching

### LINKED ISSUE #1235: Create screen for selecting the lightwalletd server
This will be reachable from the advanced settings screen (#1234).

### PR DESCRIPTION
**Pull request by HonzaR**
- Closes #1235
- Note that failures and server saving success are reported to the UI with the Android system Snackbar or via the existing Textfield’s error bottom text until we define popup design as filed in #1242
- Changelog update

https://github.com/Electric-Coin-Company/zashi-android/assets/9017390/c09462aa-e8c3-4e5a-90c7-b2df4a19feb9

> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [x] Update **documentation** as appropriate (e.g [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), [**CHANGELOG.md**](../blob/main/CHANGELOG.md), etc.)
- [x] **Run the app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

> **Note**
> It is good practice to provide before and after UI  **screenshots** in the description of this PR. This is only applicable for changes that modify the UI.

# Reviewer

- [ ] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [ ] Review the **documentation**, [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), etc. as appropriate
- [ ] **Run the app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/chooseserver/AvailableServerProvider.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// TODO [#1273]: https://github.com/Electric-Coin-Company/zashi-android/issues/1273
+
+object AvailableServerProvider {
+    // North America  | [na.lightwalletd.com](http://na.lightwalletd.com/) | 443
+    // South America | [sa.lightwalletd.com](http://sa.lightwalletd.com/) | 443
+    // Europe & Africa | [eu.lightwalletd.com](http://eu.lightwalletd.com/) | 443
+    // Asia & Oceania | [ai.lightwalletd.com](http://ai.lightwalletd.com/) | 443
+    // Plus current network defaults:
+    // Mainnet: mainnet.lightwalletd.com | 9067
+    // Testnet: lightwalletd.testnet.electriccoin.co | 9067
```
Comment (LukasKorba) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/chooseserver/AvailableServerProvider.kt:19]: [non-blocker] Are these comments needed? Everything can be easily determined from the code bellow it.

**Review by LukasKorba — CHANGES_REQUESTED** [review]
One non-blocker in the code commented. I see some inconsistencies in the video though. I don't know if the video is outdated and the behaviours changed a bit since the draft. I want @true-jared to decide the next step with this PR. The video observations:

- error handling, the red label is not in the design, consder modal error dialog
- when the custom option is selected, it should always stay selected but I see that custom option changed unexpectadle to predefined option (when a user leaves and re-enters the screen)
- Thing I don't understand, it's clear from the video that landscape is enabled. This may be just dev version but I wonder if @HonzaR spends time optimizing the layout for landscape, it would one of the thing that could optimize dev time.
- When save is pressed, iOS disables back button so it's not possible to leve the screen while processing the request. It's also greyed out so it's obvious the back button is disabled.

**Comment by LukasKorba** [comment]
Seems like one of the points has been resolved in a different [PR](https://github.com/Electric-Coin-Company/zashi-android/pull/1277)

**Comment by HonzaR** [comment]
> > when the custom option is selected, it should always stay selected but I see that custom option changed unexpectadle to predefined option (when a user leaves and re-enters the screen)
> 
> I agree with this 100%, @nuttycom was mentioning this as a hard requirement, we should never change the server without an action from the user. @HonzaR Please make sure this is the case.
> 

Let's file a follow-up on this edge case so we're not blocked at it, and the PR size does not grow. Thanks.

> > When save is pressed, iOS disables the back button, so it's impossible to leve the screen while processing the request. It's also greyed out so it's obvious the back button is disabled.
> 
> I think this is a good UX practice, we should follow on both platforms. @HonzaR Do you agree?
> 

The UI back button and system navigation are both disabled in the validation running state. The button is not greyed out yet, as design document does not provide such information. Let's file this too.

> > Thing I don't understand, it's clear from the video that landscape is enabled. This may be just dev version but I wonder if @HonzaR spends time optimizing the layout for landscape, it would one of the thing that could optimize dev time.
> 

I already explained this internally to @LukasKorba. Briefly again: rotation is one of the possible configuration changes that may happen on the device. Another could be system font size change, locale change, etc. (see more in [documentation](https://developer.android.com/guide/topics/resources/runtime-changes)), and I believe that it's the same for the iOS platform. The rotation in the video above intentionally demonstrates that the screen logic correctly handles any configuration change and does not lose its state, as that would be a rookie mistake :) As you can also see in the video, the landscape layout is the default one - rotated portrait layout.

**Review by LukasKorba — APPROVED** [review]

**Comment by true-jared** [comment]
I'm a little concerned that we are creating a trail of dependencies behind by merging the PRs as they are without implementing the requested changes. It's 2 additional issues from the last PR and 2 more from this one. I'm fine doing it when requirements are unclear but it should NOT happen in case of hard requirements. @daira as the Head of Engineering, what would you propose here and how shall we handle this going forward?

**Comment by HonzaR** [comment]
> I'm a little concerned that we are creating a trail of dependencies behind by merging the PRs as they are without implementing the requested changes. It's 2 additional issues from the last PR and 2 more from this one. I'm fine doing it when requirements are unclear but it should NOT happen in case of hard requirements. @daira as the Head of Engineering, what would you propose here and how shall we handle this going forward?

Merging PR does not mean releasing to testers in our current CI configuration. We can approach doing the follow-up any time we decide, even right after this. The intention of the video was to show any edge case we can meet, open discussion and choose the correct approach. That's the common way if we do not have such a thing defined in the limited time.

**Pull request merged**